modules/catalog/README.md

```diff
@@ -1,7 +1,54 @@
+Catalog is a component that responsible for managing descriptors of objects available in cluster,
+as well as serving as single source of truth for other components to acquire an actual state of
+the schema.
 
-See [CatalogService](src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java)
+This module provides implementation of catalog service as well as internal API for accessing the
+objects descriptors.
 
-TODO: IGNITE-19082 Add comprehensive description.
\ No newline at end of file
+## Base components description
+
+* [CatalogObjectDescriptor](src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogObjectDescriptor.java) --
+  base class for objects managed by catalog, like tables, indexes, zones, etc.
+* [CatalogService](src/main/java/org/apache/ignite/internal/catalog/CatalogService.java) -- provides
+  methods to access catalog's objects' descriptors of exact version and/or last actual version at 
+  given timestamp, which is logical point-in-time.
+* [CatalogCommand](src/main/java/org/apache/ignite/internal/catalog/CatalogCommand.java) -- denotes
+  particular modification of the catalog, like creation of particular table, for example.
+* [CatalogManager](src/main/java/org/apache/ignite/internal/catalog/CatalogManager.java) -- provides
+  methods for object manipulation (like creation of new object and/or modification of existing ones),
+  also takes care of component lifecycle.
+* [UpdateEntry](src/main/java/org/apache/ignite/internal/catalog/storage/UpdateEntry.java) -- 
+  result of applying [CatalogCommand](src/main/java/org/apache/ignite/internal/catalog/CatalogCommand.java),
+  represents delta required to move current version of catalog to state defined in the given command. 
+* [UpdateLog](src/main/java/org/apache/ignite/internal/catalog/storage/UpdateLog.java) -- distributed
+  log of incremental updates.
+
+## How it works
+
+### CatalogManager
+
+Every modification of catalog is split on two phases: accept and apply. During accept phase, given command
+is validated, then used to generate a list of update entries to save, and, finally, list is saved to 
+distributed update log. Below is a sequence diagram describing accept phase:
+![Accepting catalog update](tech-notes/accept_update_flow.png)
+
+After update entries are saved to the log, it is the job of the update log to propagate updates across the
+cluster. On every node, update log notifies catalog manager about new update entries, and latter applies
+them and stores new version of a catalog in a local cache. Below is a sequence diagram describing apply phase:
+![Applying catalog update](tech-notes/apply_update_flow.png)
+
+### UpdateLog
+
+Current implementation of update log based on a metastorage. Update entries of version N are stored by
+`catalog.update.{N}` key. Also, the latest known version is stored by `catalog.version` key. Updates
+are saved on CAS manner with condition `newVersion == value(catalog.version)`.
```

Comment:
Maybe, mentioning particular metastorage keys name is too detailed. It's up to you to keep it as is or not.

I expect to see some guatantees here, like command execition method return a future with completes, when a catalog manager got it's own update, + schema sync integration.
So, command initiator will see expected/correct catalog state after future finished.

Reply:
> I expect to see some guatantees here

good point! added `Guarantee` section

> mentioning particular metastorage keys name is too detailed

perhaps, you are right. But since you don't mind to leave it as is, I would prefer to go this way. We always can get rid of this section later if it appears too expensive to keep this up to date